### PR TITLE
[ray client] enable ray.get with >2 sec timeout (#21883)

### LINKED ISSUE #21883: [Bug] Hard-coded timeout for get()
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

I am testing ray around plasma allocating host memory of 10^3, 10^4,... 10^9 bytes, and the test was stuck on 10^9 (1 GB) with repeated retry messages.

```
2022-01-24 19:04:09,640 DEBUG worker.py:364 -- Internal retry for get [ClientObjectRef(00310a0f0a45af5cffffffffffffffffffffffff0100000002000000)]
```

raylet.out:
```
...
[2022-01-24 17:54:45,890 I 223 223] store_runner.cc:30: Allowing the Plasma store to use up to 4.7744GB of memory.
...
[2022-01-24 18:03:50,253 I 223 223] object_buffer_pool.cc:146: Not enough memory to create requested object 00310a0f0a45af5cffffffffffffffffffffffff0100000002000000, aborting
```

The raylet repeatedly attempts to create objects but it kept failing. The object store had 4GB but it says the memory space is not enough for 1GB memory allocation. Increasing/decreasing the size of object store memory didn't resolve this (the same messages are shown and spill didn't work).

~I also investigated this issue further and found that the above error happened only when I set `@ray.remote(num_gpus=1)` although I didn't touch GPU memory.~ (See my comment below)

I think this is very strange behavior and so I report this here. To be honest, I am new to ray, but do I miss something?

### Versions / Dependencies

OpenShift 4.8 (Red Hat Enterprise Linux release 8.4 (Ootpa)) with CUDA 11.4 and driver 470.82.01.
Tested images: rayproject/ray:1.7.0-py37, rayproject/ray:latest-py37 (1.9.2), rayproject/ray:nightly-py37-cu111 (2.0.0.dev0 on Jan 25).
Each cluster node has 112 CPU cores (Intel(R) Xeon(R) Gold 6258R), 1TiB memory, 4 Tesla V100-PCIE-16GB.
Huge pages are disabled.

My YAML file is like the following (I believe I do nothing special).
```
apiVersion: cluster.ray.io/v1
kind: RayCluster
metadata:
  name: ray-cluster-gpu
  namespace: ray-operator-system
spec:
  # The maximum number of workers nodes to launch in addition to the head node.
  maxWorkers: 1
  upscalingSpeed: 0
  idleTimeoutMinutes: 99999
  headPodType: head-node
  podTypes:
  - name: head-node
    # Minimum number of Ray workers of this Pod type.
    minWorkers: 0
    # Maximum number of Ray workers of this Pod type. Takes precedence over minWorkers.
    maxWorkers: 0
    rayResources: {"GPU": 0}
    podConfig:
      apiVersion: v1
      kind: Pod
      metadata:
        # Automatically generates a name for the pod with this prefix.
        generateName: head-
      spec:
        restartPolicy: Never
        volumes:
        - name: dshm
          emptyDir:
            medium: Memory
        containers:
        - name: ray-node
          imagePullPolicy: Always
          image: rayproject/ray:1.7.0-py37
          command: ["/bin/bash", "-c", "--"]
          args: ['trap : TERM INT; sleep infinity & wait;']
          ports:
          - containerPort: 6379  # Redis port
          - containerPort: 10001  # Used by Ray Client
          - containerPort: 8265  # Used by Ray Dashboard

          volumeMounts:
          - mountPath: /dev/shm
            name: dshm
          resources:
            requests:
              cpu: "1"
              memory: "16G"
              nvidia.com/gpu: "0"
            limits:
              cpu: "1"
              memory: "16G"
              nvidia.com/gpu: "0"
  - name: worker-node
    minWorkers: 1
    maxWorkers: 1
    podConfig:
      apiVersion: v1
      kind: Pod
      metadata:
        generateName: worker-
      spec:
        restartPolicy: Never
        volumes:
        - name: dshm
          emptyDir:
            medium: Memory
        containers:
        - name: ray-node
          imagePullPolicy: Always
          image: rayproject/ray:1.7.0-py37
          command: ["/bin/bash", "-c", "--"]
          args: ["trap : TERM INT; sleep infinity & wait;"]
          volumeMounts:
          - mountPath: /dev/shm
            name: dshm
          resources:
            requests:
              cpu: "8"
              memory: "16G"
              nvidia.com/gpu: "1"
            limits:
              cpu: "8"
              memory: "16G"
              nvidia.com/gpu: "1"
  headStartRayCommands:
    - ray stop
    - ulimit -n 65536; ray start --head --no-monitor --dashboard-host 0.0.0.0
  workerStartRayCommands:
    - ray stop
    - ulimit -n 65536; ray start --address=$RAY_HEAD_IP:6379

```

### Reproduction script

```
#!/usr/bin/env python
import time, ray
@ray.remote(num_gpus=1)
def gimme_object(size):
  return ray.put(bytearray(10**size))
ray.init("ray://localhost:10001", logging_level="DEBUG")
for size in range(3,10):
  time_start = time.time()
  task = gimme_object.remote(size)
  objref = ray.get(task)
  time_pull = time.time()-time_start
  print(f"took {time_pull:.3f}s for remote to put data with size={10**size} into plasma")
  time_start = time.time()
  obj = ray.get(objref)
  time_pull = time.time()-time_start
  print(f"took {time_pull:.3f}s to pull object from plasma with size={len(obj)} bytes")
  time.sleep(1)
```

### Anything else

Logs: [logs.tar.gz](https://github.com/ray-project/ray/files/7938321/logs.tar.gz)

ray memory (Note: 10.130.4.94 is the head node's IP):
```
======== Object references status: 2022-01-24 18:56:45.340288 ========
Grouping by node address...        Sorting by object size...        Display allentries per group...


--- Summary for node address: 10.130.4.94 ---
Mem Used by Objects  Local References  Pinned        Pending Tasks  Captured in Objects  Actor Handles
1000000969.0 B       4, (1000000969.0 B)  0, (0.0 B)    0, (0.0 B)     0, (0.0 B)           0, (0.0 B)   

--- Object references for node address: 10.130.4.94 ---
IP Address       PID    Type    Call Site               Size    Reference Type      Object Ref                                              
10.130.4.94      615    Driver  disabled                86.0 B  LOCAL_REFERENCE     00ffffffffffffffffffffffffffffffffffffff0100000001000000

10.130.4.94      615    Driver  disabled                264.0 B  LOCAL_REFERENCE     16310a0f0a45af5cffffffffffffffffffffffff0100000001000000

10.130.4.94      615    Driver  disabled                600.0 B  LOCAL_REFERENCE     00ffffffffffffffffffffffffffffffffffffff0100000002000000

10.130.4.94      615    Driver  <unknown>               1000000019.0 B  LOCAL_REFERENCE     00310a0f0a45af5cffffffffffffffffffffffff0100000002000000

To record callsite information for each ObjectRef created, set env variable RAY_record_ref_creation_sites=1

--- Aggregate object store stats across all nodes ---
Plasma memory usage 953 MiB, 3 objects, 10.47% full, 10.47% needed
Objects consumed by Ray tasks: 0 MiB.
```

### Are you willing to submit a PR?

- [x] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by takeshi-yoshimura**
Commit 2cf4c72 ("[ray client] Fix ctrl-c for ray.get() by setting a short-server side timeout") introduced a short server-side timeout not to block later operations.

However, the fix implicitly assumes that get() is complete within MAX_BLOCKING_OPERATION_TIME_S (two seconds). This becomes a problem when apps use heavy objects or limited network I/O bandwidth that require more than two seconds to push all chunks. The current retry logic needs to re-push from the beginning of chunks and block clients with the infinite re-push.

I updated the logic to directly pass timeout if it is explicitly given. Without timeout, it still uses MAX_BLOCKING_OPERATION_TIME_S for polling with the short server-side timeout.

Initially, I thought the fix should be exponential back-off, but that may change the behavior of many apps. So, this patch focuses on enabling bigger timeouts.

## Related issue number

Closes #21883

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ckw017 — APPROVED** [review]
Looks good to me, can you rebase/merge with master? That should rerun CI and hopefully fix the failing tests.

**Comment by takeshi-yoshimura** [comment]
@ckw017 
Thanks. I merged with master.

**Review by ckw017 — CHANGES_REQUESTED** [review]
Hmm, re-evaluating this, I think this overloads the meaning of `ray.get`'s timeout a bit too much. For example, users will have to explicitly add timeout to work in client, even though it would be unnecessary when attaching a driver directly.

@takeshi-yoshimura would being able to set an environment variable, i.e. something like `RAY_CLIENT_MAX_BLOCKING_OPERATION_TIME_S=20` to increase the timeout work for your usecase instead?

Alternatively we can try bumping `MAX_BLOCKING_OPERATION_TIME` to a higher value

cc @ericl (author of #14425)

**Comment by ckw017** [comment]
Hmm, re-evaluating this, I think this overloads the meaning of `ray.get`'s timeout a bit too much. For example, users will have to explicitly add timeout to work in client, even though it would be unnecessary when attaching a driver directly.

@takeshi-yoshimura would being able to set an environment variable, i.e. something like `RAY_CLIENT_MAX_BLOCKING_OPERATION_TIME_S=20` to increase the timeout work for your usecase instead?

cc @ericl (author of #14425)

**Comment by takeshi-yoshimura** [comment]
@ckw017 
Updated the code to introduce a new environmental variable `RAY_CLIENT_MAX_BLOCKING_OPERATION_TIME_S` according to your request.

File: python/ray/util/client/worker.py [hunk 1]
```diff
@@ -421,14 +421,17 @@ def get(self, vals, *, timeout: Optional[float] = None) -> Any:
         else:
             deadline = time.monotonic() + timeout
 
+        max_blocking_operation_time = MAX_BLOCKING_OPERATION_TIME_S
```
Comment (ckw017) [inline python/ray/util/client/worker.py:424]:
```suggestion
        if 'RAY_CLIENT_MAX_BLOCKING_OPERATION_TIME_S' in os.environ:
            max_blocking_operation_time = float(os.environ['RAY_CLIENT_MAX_BLOCKING_OPERATION_TIME_S'])
```

**Comment by ckw017** [comment]
Looks good, looks like the only thing failing is the [lint](https://buildkite.com/ray-project/ray-builders-pr/builds/28819#bac99b5f-5699-44cc-84b9-05de241c720a/1492-1578). Can you update to match what the linter is suggesting?

Screen Shot 2022-04-05 at 9 59 15 AM

**Comment by takeshi-yoshimura** [comment]
I fixed the format. thanks.

**Review by ckw017 — APPROVED** [review]

**Comment by ckw017** [comment]
Looks like we had a few problems with our testing infra when you first ran the PR, if you merge with master and push again it should be good to go!

**Comment by takeshi-yoshimura** [comment]
Merged again. let's see...

**Comment by ckw017** [comment]
Okay, looks like failures are:
* LinkCheck (safe to ignore, failing on master as well)
* rllib:learning_tests_pendulum_ddppo (safe to ignore, failing on master)
* rllib:learning_tests_cartpole_ddppo (safe to ignore, failing on master at time of merge)
* mac tests (Looks like something in our infra dropped a bunch of mac agents around that time, IMO safe to ignore)

**Comment by ckw017** [comment]
cc @ijrsvt can you take a look/merge this when you get the chance

**Review by ijrsvt — APPROVED** [review]
LGTM

**Comment by ijrsvt** [comment]
RLLIB Tests & OSX `test_shuffle` are failing on master at the point of rebase.

**Pull request merged**